src/log.rs

```diff
@@ -0,0 +1,396 @@
+// Copyright 2020, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use glib_sys;
+use once_cell::sync::Lazy;
+use std::boxed::Box as Box_;
+use std::sync::Mutex;
+use translate::*;
+use GString;
+use LogLevelFlags;
+use LogWriterOutput;
+
+pub enum LogValue {
+    String(String),
+    Value(Vec<u8>),
+}
+
+pub struct LogField {
+    key: GString,
+    value: LogValue,
+}
+
+#[doc(hidden)]
+impl ToGlib for LogField {
+    type GlibType = glib_sys::GLogField;
+
+    fn to_glib(&self) -> glib_sys::GLogField {
+        match self.value {
+            LogValue::String(ref s) => {
+                let s: *const libc::c_char = s.to_glib_none().0;
+                glib_sys::GLogField {
+                    key: self.key.to_glib_none().0,
+                    value: s as _,
+                    length: -1,
+                }
+            }
+            LogValue::Value(ref v) => glib_sys::GLogField {
+                key: self.key.to_glib_none().0,
```

Comment:
Not fully sure that this works fine to, we need more comment in GString implementation.